### ISSUE TITLE
rbenv install 2.2.10 fails in Ubuntu 20.04 - Need OpenSSL 1.0.0, how to have multiple openssl versions?

### ISSUE THREAD

**Issue by cratag** [rbenv/ruby-build#1740]
According to docs:
>When building Ruby 2.3 or older, you must ensure that OpenSSL 1.0 is available somewhere on the system and point to it via the --with-openssl-dir configure flag. At the moment, most Linux distributions that come with OpenSSL 1.1 have a separate package for OpenSSL 1.0.

`openssl version -v` returns `OpenSSL 1.1.1f 31 Mar 2020`
I've installed 1.0.2 [following this](https://askubuntu.com/questions/1000629/how-to-install-openssl-1-0-2-with-default-openssl-1-1-1-on-ubuntu-16-04), but then I only had `OpenSSL 1.0.2l 25 May 2017` which interferes with another 3.0.0 Ruby App I have.

Ubuntu has no 'brew'. I installed brew following [their docs](https://brew.sh/).

Then this command
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.0)" rbenv install 2.2.10`
returns

`Error: No available formula with the name "openssl@1.0".`

so I followed [this link](https://github.com/rbenv/homebrew-tap/issues/1)
and ran:
`brew tap rbenv/tap`
`brew install rbenv/tap/openssl@1.0`

and then 
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.0)" rbenv install 2.2.10`

> result:
> BUILD FAILED (Ubuntu 20.04 using ruby-build 20210309)
> 
> Inspect or clean up the working tree at /tmp/ruby-build.20210322123729.48434.WmdkHX
> Results logged to /tmp/ruby-build.20210322123729.48434.log
> 
> Last 10 log lines:
> make[1]: *** Waiting for unfinished jobs....
> linking shared-object psych.so
> installing default psych libraries
> make[2]: Leaving directory '/tmp/ruby-build.20210322123729.48434.WmdkHX/ruby-2.2.10/ext/psych'
> linking shared-object pathname.so
> make[2]: Leaving directory '/tmp/ruby-build.20210322123729.48434.WmdkHX/ruby-2.2.10/ext/pathname'
> linking shared-object nkf.so
> make[2]: Leaving directory '/tmp/ruby-build.20210322123729.48434.WmdkHX/ruby-2.2.10/ext/nkf'
> make[1]: Leaving directory '/tmp/ruby-build.20210322123729.48434.WmdkHX/ruby-2.2.10'
> make: *** [uncommon.mk:189: build-ext] Error 2

Maybe there's an easier command without brew for linux?

**Comment by cratag** [comment]
[This ](https://gist.github.com/souzagab/0ae60e61939d51385de87904b65b2da2)was the solution I found, and with a small tweak it worked.
Thanks to EliKazap
Edit /`etc/apt/sources.list` and add this at the end of the file: `deb http://security.ubuntu.com/ubuntu bionic-security main`

then

    Run sudo apt-get update && apt-cache policy libssl1.0-dev

    Run sudo apt-get install libssl1.0-dev


check if the version is correct with `./usr/local/ssl/bin/openssl version`

and then

    RUBY_CONFIGURE_OPTS="--with-openssl-dir='/usr/local/ssl/bin'" rbenv install 2.2.10

**Comment by ibariens** [comment]
Thanks @cratag for also taking the time to post your solution. That helped me a lot.
In case it helps anyone, I used your solution but I built the package instead of adding an item to sources.list

1. Ensure you have 'curl' and 'make' packages installed
2. Download and unzip openssl
`- curl https://www.openssl.org/source/openssl-1.0.2l.tar.gz | tar xz && cd openssl-1.0.2l`
3. install openssl to directory /old-openssl-version-1.0.2
```
- ./config --prefix=/old-openssl-version-1.0.2 --openssldir=/old-openssl-version-1.0.2
- sudo make
- sudo make install
```
4. use the 1.0 openssl version for building ruby
` RUBY_CONFIGURE_OPTS="--with-openssl-dir='/old-openssl-version-1.0.2/bin'" rbenv install 2.2.10 `
5. Optional, install older version of bundler
`gem install bundler -v '~>1'`

**Comment by jaderiver62** [comment]
Hi folks,
I ran into the same issue and I tried both @cratag and @ibariens  solutions and I get the following error:

> 
> Hit:1 https://deb.nodesource.com/node_12.x hirsute InRelease
> Get:2 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
> Hit:3 https://dl.google.com/linux/chrome/deb stable InRelease                   
> Hit:4 http://archive.ubuntu.com/ubuntu impish InRelease                         
> Hit:5 http://archive.ubuntu.com/ubuntu impish-updates InRelease
> Hit:6 http://archive.ubuntu.com/ubuntu impish-backports InRelease
> Err:2 http://security.ubuntu.com/ubuntu bionic-security InRelease
>   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
> Hit:7 http://archive.ubuntu.com/ubuntu impish-security InRelease
> Reading package lists... Done
> W: GPG error: http://security.ubuntu.com/ubuntu bionic-security InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
> E: The repository 'http://security.ubuntu.com/ubuntu bionic-security InRelease' is not signed.
> N: Updating from such a repository can't be done securely, and is therefore disabled by default.
> N: See apt-secure(8) manpage for repository creation and user configuration details.
> 

Any help is greatly appreciated!

**Comment by ctrlcctrlv** [comment]
Just a note for other Googlers:

In Bash, this won't work—

```bash
RUBY_CONFIGURE_OPTS="--with-openssl-dir='/usr/local/ssl/bin'" rbenv install 2.2.10
```

The single quotes are taken literally.

You need:

```bash
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=/usr/local/ssl/bin"
```

Or if following good recommended instructions at https://github.com/rbenv/ruby-build/discussions/1940#discussioncomment-2663209:

```bash
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=/opt/openssl-1.1.1o" 
```
